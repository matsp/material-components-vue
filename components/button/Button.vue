<template>
  <a
    v-if="href"
    :class="classes"
    :href="href"
    v-bind="$attrs"
    role="button"
    class="mdc-button"
    v-on="$listeners"
  >
    <slot name="icon" />
    <slot />
  </a>
  <button
    v-else
    :class="classes"
    v-bind="$attrs"
    class="mdc-button"
    v-on="$listeners"
  >
    <slot name="icon" />
    <slot />
  </button>
</template>

<script>
import { MDCRipple } from '@material/ripple'
import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  props: {
    raised: {
      type: Boolean,
      default: false
    },
    unelevated: {
      type: Boolean,
      default: false
    },
    outlined: {
      type: Boolean,
      default: false
    },
    dense: {
      type: Boolean,
      default: false
    },
    href: {
      type: String,
      default: ''
    }
  },
  data () {
    return {
      mdcRipple: undefined,
      slotObserver: undefined
    }
  },
  computed: {
    classes () {
      return {
        'mdc-button--raised': this.raised,
        'mdc-button--unelevated': this.unelevated,
        'mdc-button--outlined': this.outlined,
        'mdc-button--dense': this.dense
      }
    }
  },
  watch: {
    classes () {
      this.mdcRipple.destroy()
      this.mdcRipple = MDCRipple.attachTo(this.$el)
    }
  },
  mounted () {
    this.updateSlot()
    this.slotObserver = new MutationObserver(() => this.updateSlot())
    this.slotObserver.observe(this.$el, {
      childList: true,
      subtree: true
    })
    this.mdcRipple = MDCRipple.attachTo(this.$el)
  },
  beforeDestroy () {
    this.slotObserver.disconnect()
    if (typeof this.mdcRipple !== 'undefined') {
      this.mdcRipple.destroy()
    }
  },
  methods: {
    updateSlot () {
      if (this.$slots.icon) {
        this.$slots.icon.map(n => {
          n.elm.classList.add('mdc-button__icon')
          n.elm.setAttribute('aria-hidden', 'true')
        })
      }
    }
  }
}
</script>
