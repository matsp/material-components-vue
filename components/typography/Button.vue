<template>
  <span class="mdc-typography--button">
    <slot />
  </span>
</template>

<script>
import { baseComponentMixin, themeClassMixin } from '@components/base'

export default {
  mixins: [baseComponentMixin, themeClassMixin]
}
</script>
