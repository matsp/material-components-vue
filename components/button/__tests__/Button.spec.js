import 'mutationobserver-shim'
import { mount } from '@vue/test-utils'
import Button from '../Button.vue'

describe('Button', () => {
  it('should mount', () => {
    const wrapper = mount(Button)
    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.vm.$data.mdcRipple).toBeDefined()
    expect(wrapper.vm.$data.slotObserver).toBeDefined()
  })

  it('should render with no prop', () => {
    const wrapper = mount(Button)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-button')
  })

  it('should render as raised', () => {
    const wrapper = mount(Button, {
      propsData: {
        raised: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-button--raised')
  })

  it('should render as unelevated', () => {
    const wrapper = mount(Button, {
      propsData: {
        unelevated: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-button--unelevated')
  })

  it('should render as outlined', () => {
    const wrapper = mount(Button, {
      propsData: {
        outlined: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-button--outlined')
  })

  it('should render as dense', () => {
    const wrapper = mount(Button, {
      propsData: {
        dense: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-button--dense')
  })

  it('should render without ripple', () => {
    const wrapper = mount(Button, {
      propsData: {
        ripple: false
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.vm.$data.mdcRipple).toBeUndefined()
  })
})
