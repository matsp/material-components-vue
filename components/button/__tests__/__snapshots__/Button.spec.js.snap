// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render as dense 1`] = `<button class="mdc-button mdc-button--dense"> <span class="mdc-button__label"></span> </button>`;

exports[`Button should render as outlined 1`] = `<button class="mdc-button mdc-button--outlined"> <span class="mdc-button__label"></span> </button>`;

exports[`Button should render as raised 1`] = `<button class="mdc-button mdc-button--raised"> <span class="mdc-button__label"></span> </button>`;

exports[`Button should render as unelevated 1`] = `<button class="mdc-button mdc-button--unelevated"> <span class="mdc-button__label"></span> </button>`;

exports[`Button should render with no prop 1`] = `<button class="mdc-button"> <span class="mdc-button__label"></span> </button>`;

exports[`Button should render without ripple 1`] = `<button class="mdc-button"> <span class="mdc-button__label"></span> </button>`;
